[2783] printf command

elee1766 opened the pull request:
address #2783

still wip.


features:
[x] single pass
[x] single write to stdout
[x] errors to stderr
[x] parses flags in  %[flags][width][.precision][length]specifier format, with support for *
[x] \" \\ \a \b \c \e \f \n \r \t \v \NNN \0NNN \xHH \uHHHH \uHHHHHHHH
[x] %b
[x] %s
[x] %q
[-] %d partially implemented, missing width, precision, length, padding, etc. can parse hex strings. 
[ ] {iouxXfeEgGc}

elee1766 commented:
@rminnich hi

i had some thoughts while implementation, thought you would be a good person to give opinions, sorry if you are not the right person to ping for this 

first of all,  each implementation seems to deals with printf differently. they not only have different features for arguments, but also support different escape codes, etc. 

for instance, 
`printf "%d" "123+123"` is valid in only `zsh` , because `zsh` supports math expressions
`printf \u1230` is not supported by debian `sh` and simply returns `\u1230` 
`printf \u123` errors in coreutils, pads to `\u0123` in bash & zsh
`printf --help` simply returns `--help` in `zsh`, while returning help menu for bash & coreutils
`printf %q` is different for every shell, because it formats in a way that is friendly to the shell that it is in
the list continues on and on

as a result, i think the best way to do printf would be...

1. move cmd/printf to pkg/printf
2. create cmd/printf which mirrors the behavior of coreutils printf
3. create printf builtin in `gosh`, which has format code %q/%Q for gosh semantics, and can add any shell specific integrations

i'm not too familiar with gosh,  if it is fully posix compliant, then we can probably share %q formatters, and just need to make sure that input args get properly encoded/passed in a friendly manner

rminnich commented on cmds/core/printf/printf.go:
```diff
@@ -0,0 +1,114 @@
+
+type printf struct {
+	Args   []string
+	Stdout io.Writer
```
since you have a new, is there a reason to export these? Does some package require it? I don't know.

elee1766 replied:
nope it was just to make clear for me which were constructor args and which weren't while I was writing

rminnich commented on cmds/core/printf/printf.go:
```diff
@@ -0,0 +1,114 @@
+	"os"
+)
+
+type printf struct {
```
another thing I don't know. Does this type implement some interface? If it does, even though lint does not require it, a comment might be helpful for those who come later.

elee1766 replied:
no, but maybe it should. 

i thought briefly about making printf an io.Reader instead of accepting an io.Writer as input, but that felt wrong.

rminnich commented on cmds/core/printf/printf.go:
```diff
@@ -0,0 +1,114 @@
+	arguments []string
+}
+
+func NewPrinter(stdout, stderr io.Writer, args []string) *printf {
```
// NewPrinter returns a printf

or some kind of comment.

rminnich commented on cmds/core/printf/printf.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+func (c *printf) run() {
+	if len(c.Args) < 1 {
```
The command way to handle this would be, in the New function, to have an Arg0 and Args.
Or, further, set  c.format in the New. By the time printf is called, this kind of arg checking ought to have been done, and New is a good place for it.

rminnich commented on cmds/core/printf/printf.go:
```diff
@@ -0,0 +1,114 @@
+		c.Stderr.Write([]byte("printf: not enough arguments\n"))
+		return
+	}
+	c.format = c.Args[0]
```
This could be done in the New

rminnich commented on cmds/core/printf/printf.go:
```diff
@@ -0,0 +1,114 @@
+
+func (c *printf) run() {
+	if len(c.Args) < 1 {
+		c.Stderr.Write([]byte("printf: not enough arguments\n"))
```
Would be nice if this could return an error, not do this write.

rminnich commented on cmds/core/printf/printf.go:
```diff
@@ -0,0 +1,114 @@
+		return
+	}
+	c.format = c.Args[0]
+	if len(c.Args) > 1 {
```
This could be set in New.

rminnich commented on cmds/core/printf/printf.go:
```diff
@@ -0,0 +1,114 @@
+	w := new(bytes.Buffer)
+	err := c.exec(w)
+	if err != nil {
+		c.Stderr.Write([]byte("printf: " + err.Error() + "\n"))
```
This can return an error, it makes tests easier. Also, you can use fmt.Errorf, with a wrapped error type (%w) which makes your tests simpler.

rminnich commented on cmds/core/printf/printf_test.go:
```diff
@@ -0,0 +1,70 @@
+	type testCase struct {
+		a   []string
+		e   string
+		err string
```
if you have run return an error, you can use wrapper errors, and the err here can be an error type. That is preferred for tests now.

rminnich replied:
you can use errors.Is.

rminnich commented on cmds/core/printf/utils.go:
```diff
@@ -0,0 +1,211 @@
+	"strings"
+)
+
+func interpret(
```
in general, in Go, this kind of arg formatting is not really idiomatic. Please fit this declaration on one line.

rminnich commented on cmds/core/printf/utils.go:
```diff
@@ -0,0 +1,211 @@
+				fmt.Fprintf(o, "%s", arg)
+				continue
+			default:
+				return fmt.Errorf("%s: %s", "%"+string(n), "invalid directive")
```
if you can make an error: 
var errIInvalidDirective = errors.New("invalid directive")
and then in the fmt.Errorf, wrap it with %w, then your tests are easier to write.

rminnich commented on cmds/core/printf/utils.go:
```diff
@@ -0,0 +1,211 @@
+	"strings"
+)
+
+func interpret(
```
I'm a bit surprised to see you had to write your own interpreter, is the one in fmt not usable for some reason?

elee1766 replied:
the fmt in golang is semantically different from every implementation that I compared with (coreutils, bash, zsh, sh) 

I'm pretty sure this will be inevitable, especially if we need to implement format code %q

rminnich commented on cmds/core/printf/utils.go:
```diff
@@ -0,0 +1,211 @@
+	return nil
+}
+
+func readOctal(fr *strings.Reader, o *bytes.Buffer) {
```
I am fairly certain you can use strconv package for this, just set the base.

elee1766 replied:
I do within readoctal, but we need to be able to stop reading the octal after either N characters OR upon the first non decimal character, which I can't do with just strconv

rminnich commented on cmds/core/printf/utils.go:
```diff
@@ -0,0 +1,211 @@
+	if len(hexcode) == 0 {
+		o.WriteRune('\\')
+	} else {
+		i, err := strconv.ParseInt(hexcode, 16, 32)
```
I'm not sure I see the need for this. Am I missing something (likely :-)?

elee1766 replied:
do you mean the error check? I don't think the error check is needed here, I'm not sure if it's better to add a comment explaining why, or just handle it

elee1766 replied:
oh do you mean the check if hexcode is 0? i think that bit is wrong actually :-)

elee1766 commented:
@rminnich i've changed the error handling, moved printf to another package, and added a few comments. 

as to reasons why i had to start on a custom interpreter instead of go fmt, here are a few - 
1. some format codes, e.g. %q, %b have completely different meanings
2. some format codes, e.g. %x, %X, %p have slightly different rules 
3. go is missing format code %i
4. go is missing escape sequence \e

along with that, parsing arguments is completely different. for instance, `printf '%d' 0x1234' interprets the hex and will output `4660` and `printf '%s' 0x1234` will keep 0x1234 as a string, which go's fmt can't do. 

basically, there will actually need to be helper functions which parse the parameters into their correct form based off the format code they match with

after that though, we can reuse some logic and use fprintf with the go directive, as i am with %s

elee1766 commented:
ok, i've created a framework to create the rest of the format codes.

rminnich commented:
I think this person is not coming back.
